### PR TITLE
Add IHttpUpgradeFeature to TestServer for SignalR WebSocket support

### PR DESCRIPTION
**Pull request by BrennanConroy**
SignalR checks for `IHttpWebSocketFeature` which is set by the WebSocket middleware if the `IHttpUpgradeFeature` exists.

https://github.com/dotnet/aspnetcore/blob/4fc4748c196de4bae8da1e146638f484e7e4b99c/src/SignalR/common/Http.Connections/src/Internal/HttpConnectionDispatcher.cs#L377-L380

I don't know how much validation we need if any for saying that the request supports websockets. Also, currently throwing from `IHttpUpgradeFeature.UpgradeAsync` since TestServer doesn't use that for the stream.

Going to see if we can get this into preview6 since the new `WebSocketFactory` option was added to SignalR in preview5 with the intent to allow TestServer websockets to work with the client.

### CODE REVIEW

**Review by shirhatti — APPROVED** [review]
Thanks Brennan!

File: src/Hosting/TestHost/src/UpgradeFeature.cs [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+using System;
+using System.IO;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http.Features;
+
+namespace Microsoft.AspNetCore.TestHost
+{
+    internal class UpgradeFeature : IHttpUpgradeFeature
+    {
+        public bool IsUpgradableRequest => true;
```
Comment (Tratcher) [inline src/Hosting/TestHost/src/UpgradeFeature.cs:13]: The presence of the feature should be enough, upgrades do not need to be possible.
```suggestion
        public bool IsUpgradableRequest => false;
```

File: src/Hosting/TestHost/src/UpgradeFeature.cs [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+        // TestHost provides an IHttpWebSocketFeature so it wont call UpgradeAsync()
+        public Task<Stream> UpgradeAsync()
+        {
+            throw new NotImplementedException();
```
Comment (Tratcher) [inline src/Hosting/TestHost/src/UpgradeFeature.cs:18]:
```suggestion
            throw new NotSupportedException();
```

File: src/Hosting/TestHost/test/WebSocketClientTests.cs [hunk 3]
```diff
@@ -54,5 +55,48 @@ public async Task ConnectAsync_ShouldSetRequestProperties(string requestUri, str
             Assert.Equal(expectedHost, capturedHost);
             Assert.Equal("/connect", capturedPath);
         }
+
+        [Fact]
+        public async Task CanAcceptWebSocket()
```
Comment (Tratcher) [inline src/Hosting/TestHost/test/WebSocketClientTests.cs:60]: You want a different test for the normal TestServer scenario that shows the upgrade and WebSocket features are present, but an upgrade isn't possible.

**Comment by BrennanConroy** [comment]
@Pilchie for preview6 consideration

File: src/Hosting/TestHost/test/WebSocketClientTests.cs [hunk 4]
```diff
@@ -54,5 +55,77 @@ public async Task ConnectAsync_ShouldSetRequestProperties(string requestUri, str
...
+                        Assert.False(upgradeFeature.IsUpgradableRequest);
+                        await Assert.ThrowsAsync<NotSupportedException>(() => upgradeFeature.UpgradeAsync());
+
+                        var webSocketFeature = c.Features.Get<IHttpWebSocketFeature>();
+                        Assert.NotNull(webSocketFeature);
+                        Assert.True(webSocketFeature.IsWebSocketRequest);
+                    });
+                })))
+            {
+                var client = testServer.CreateWebSocketClient();
```
Comment (Tratcher) [inline src/Hosting/TestHost/test/WebSocketClientTests.cs:116]: Sorry, I meant that you'd send a non-websocket request here like you do in the signalr negotiation scenario. Including `app.UseWebSockets();` in this test makes sense and the checking that it does add the `IHttpWebSocketFeature` SignalR needs.

**Review by Tratcher — APPROVED** [review]

**Comment by Pilchie** [comment]
Can you send tactics a mail for approval at this point?  Since builds have been sent out to start validation, I don't want to reset them without approval.

**Comment by BrennanConroy** [comment]
> Since builds have been sent out to start validation, I don't want to reset them without approval.

Darn, I don't think it's worth resetting for this.

**Comment by Pilchie** [comment]
Actually, I might have confused Preview 6 with servicing builds, I don't think those are out, but most teams *are* mailing Tactics for approval.

**Comment by BrennanConroy** [comment]
@dotnet/aspnet-build Tactics approved over email this morning, merge when convenient!

**Pull request merged**